File: AnagramChecker.java
import java.util.Arrays;

// Anagram means exact the same charaters but in different orders

public class AnagramChecker {
    public static boolean isAnagram(String s, String t) {
        // Check if lengths are different, if yes, they can't be anagrams
        if (s.length() != t.length()) {
            return false;
        }

        // Use an array to count the frequency of each character
        int[] charCount = new int[26]; // Assuming only lowercase English letters

        // Count the frequency of characters in string s
        for (char c : s.toCharArray()) {
            charCount[c - 'a']++;
        }
        // After this process, the charCount array would look like this: [2, 1, ..., 0]

        // Decrement the count for characters in string t
        for (char c : t.toCharArray()) {
            charCount[c - 'a']--;

            // If a character count becomes negative, strings are not anagrams
            if (charCount[c - 'a'] < 0) {
                return false;
            }
        }

        // If the code reaches here, both strings are anagrams
        return true;
    }

    public static void main(String[] args) {
        // Example usage
        String s1 = "anagram";
        String t1 = "nagaram";
        System.out.println(isAnagram(s1, t1)); // Output: true

        String s2 = "rat";
        String t2 = "tar";
        System.out.println(isAnagram(s2, t2)); // Output: false

        String s3 = "bbb";
        String t3 = "bba"; // If any char below zero means not anagram
        System.out.println(isAnagram(s3, t3)); // Output: false
    }
}

// The time complexity of this solution is O(n), where n is the length of the
// strings. This is because we iterate through each character in both strings
// once. The space complexity is O(1) since the size of the charCount array is
// constant (26 for lowercase English letters).

File: AnagramChecker.py
# Anagram means exact the same characters but in different orders

def is_anagram(s, t):
    # Check if lengths are different, if yes, they can't be anagrams
    if len(s) != len(t):
        return False

    # Use a dictionary to count the frequency of each character
    char_count = {}

    # Count the frequency of characters in string s
    for c in s:
        char_count[c] = char_count.get(c, 0) + 1
        print(char_count)

    # Decrement the count for characters in string t
    for c in t:
        if c not in char_count:
            # If a character count becomes negative or not available, strings are not anagrams
            return False
        char_count[c] -= 1
        print('After Comparison: ', char_count)

    # If the code reaches here, both strings are anagrams
    return True

# Example usage
s1 = "anagram"
t1 = "nagaram"
print(is_anagram(s1, t1))  # Output: True

s2 = "rat"
t2 = "tar"
print(is_anagram(s2, t2))  # Output: False

s3 = "bbb"
t3 = "bba"
print(is_anagram(s3, t3))  # Output: False
